### PR TITLE
Creating an ASN with the same number and same cache group should not be allowed

### LINKED ISSUE #5020: Bug that let user create two ASNs with same number in the same cachegroup
## I'm submitting a ...

-  bug report

## Traffic Control components affected ...

-  Traffic Ops

## Current behavior:

1. Go to Traffic Portal
2. Click on Topology -> ASNs.
3. Create a ASNs with valid number.
4. Create another ASNs with same number in the same cache group like step 3. Notice that TP let user create it.
## Expected / new behavior:

TO and TP should not let user create 2 ASNs with same number in the same cachegroup

### PR DESCRIPTION
**Pull request by srijeet0406**
## What does this PR (Pull Request) do?

- [x] This PR fixes #5020 #5021  

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

Make sure the unit tests and API tests pass
Run TO and TP locally, and try to create two asns with the same number and same cachegroup ID. This should fail.
Also, try to GET an asn by specifying a non numeric ID for the cachegroup ID as the query param. You should get a 400 instead of a 500 now.

## If this is a bug fix, what versions of Traffic Control are affected?

- master

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by jhg03a** [comment]
It's also invalid to assign an ASN to more than one cachegroup

**Comment by mattjackson220** [comment]
I think it might be worth throwing a note in the ASNs docs to call this out as well? or is it a common enough rule that people will know?

**Comment by rawlinp** [comment]
@jhg03a is that actually invalid or just something that seems invalid in our environment? I could imagine an environment having the same ASN on multiple cachegroups.

**Comment by jhg03a** [comment]
It's required for us, but I think it ought to be required generally even though it's not been enforced today.  The goal of tracking the mapping of ASNs to Cachegroups is so that a transitive mapping between subnets and cachegroups can be created as part of CZF file generation.  Trying to assign the same subnet to multiple cachegroups doesn't make a ton of sense to me.  Keep in mind too that some entities don't break their subnets into geographically based ASNs which CDNs have no control over or input on.  So in this case the argument for multiple mapping could be made but it's still just guessing at that point as the CZF file used by TR will only ever use one path to the best of my knowledge.  I think that's still the right behavior as well because you want IPs to land in cachegroups by way of CZF hits consistently.  Having multiple options there could lead to indeterminate behavior also making cachegroup capacity planning more difficult.

**Comment by rawlinp** [comment]
CZF generation doesn't really have a bearing on how this API should be restricted since that is external to the API, and adding new restrictions to the API like that might violate existing data, causing the DB upgrade to fail.

**Comment by srijeet0406** [comment]
So I'm assuming that the scope of this ticket will be reverted to `Creating an ASN with the same number and same cache group should not be allowed`?

**Comment by rob05c** [comment]
Does it ever make sense to assign the same subnet to multiple cachegroups?

What will Traffic Router do if that happens? Is it ever useful to do that?

>adding new restrictions to the API like that might violate existing data, causing the DB upgrade to fail (assuming it would be a DB constraint).

Can we at least make it an API constraint to prevent new ones, if it's always wrong?

**Comment by rawlinp** [comment]
I disagree for the reasons I've already stated, but if @rob05c and @jhg03a feel strongly enough for us to add API restrictions for external business reasons, then feel free.

**Comment by rob05c** [comment]
I don't know enough about the issue to feel strongly. That's why I was asking, I honestly don't know the answer to those questions.

**Comment by jhg03a** [comment]
I do believe we should for reasons I've already stated.

**Comment by srijeet0406** [comment]
If thats the case then no more changes are required to this PR.

File: docs/source/api/v3/asns.rst [hunk 1]
```diff
@@ -106,6 +106,7 @@ Response Structure
 ``POST``
 ========
 Creates a new :abbr:`ASN (Autonomous System Number)`.
```
Comment (mattjackson220) [inline docs/source/api/v3/asns.rst:108]: needs a blank line before the note as well

File: traffic_ops/traffic_ops_golang/asn/asns_test.go [hunk 2]
```diff
@@ -134,3 +134,39 @@ func TestValidate(t *testing.T) {
...
+	reqInfo := api.APIInfo{Tx: db.MustBegin()}
+	asnNum := 2
+	cachegroupID := 10
+	asn := TOASNV11{
+		api.APIInfoImpl{&reqInfo},
+		tc.ASNNullable{ASN: &asnNum, CachegroupID: &cachegroupID},
+	}
+	err = asn.ASNExists()
+	if err == nil {
+		t.Fatalf("expected no error but got %v", err.Error())
```
Comment (mattjackson220) [inline traffic_ops/traffic_ops_golang/asn/asns_test.go:167]: this error message is flipped. should be "expected an error but got error = nil"

Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/asn/asns_test.go:167]: it's also going to segfault whenever the test fails

**Review by mattjackson220 — APPROVED** [review]
Looks good! unit tests pass, API v1, v2, and v3 tests pass, docs build and look good, manual testing works as expected.

**Pull request merged**